presto-main/src/main/java/com/facebook/presto/operator/aggregation/ReduceAggregationFunction.java

```diff
@@ -199,94 +145,52 @@ else if ((stateType.getJavaType() == Block.class)) {
                 new ParameterMetadata(INPUT_CHANNEL, stateType));
     }
 
-    public static void input(NullableLongState state, Object value, long initialStateValue, BinaryFunctionInterface inputFunction, BinaryFunctionInterface combineFunction)
+    public static void input(Type type, ReduceAggState state, Object value, Object initialStateValue, BinaryFunctionInterface inputFunction, BinaryFunctionInterface combineFunction)
     {
-        if (state.isNull()) {
-            state.setNull(false);
-            state.setLong(initialStateValue);
+        if (state.getValue() == null) {
+            state.setValue(initialStateValue);
         }
-        state.setLong((long) inputFunction.apply(state.getLong(), value));
-    }
-
-    public static void input(NullableDoubleState state, Object value, double initialStateValue, BinaryFunctionInterface inputFunction, BinaryFunctionInterface combineFunction)
-    {
-        if (state.isNull()) {
-            state.setNull(false);
-            state.setDouble(initialStateValue);
+        try {
+            state.setValue(inputFunction.apply(state.getValue(), value));
```

Comment:
nits: Logically it's clearer to have
```
Object stateValue;
try {
   stateValue = inputFunction.apply(state.getValue(), value);
}
catch (NullPointerException npe) {
  stateValue = null;
}
state.setValue(stateValue);
```

Reply:
Performance-wise I thought what I have is better

Reply:
I'd thought compiler can optimize this? No?

Reply:
Maybe but also I actually think my version is more readable set the value to null when NPE.